### PR TITLE
Hide password form fields by default

### PR DESCRIPTION
**Pull request by daniel-beck**
The goal here is twofold:

* To prevent password auto-fill
* To reduce the times browsers will offer to save (unrelated) passwords that are part of form submissions.

## Example: Proxy Configuration

<details>
<summary>When the field value is empty to begin with, show a <i>text</i> field to prevent auto-fill.</summary>

proxy-empty
</details>

<details>
<summary>When the user enters a value, the field type is switched to password to mask it.</summary>

proxy-empty-filled
</details>

<details>
<summary>When a value is already filled in, show just a placeholder:</summary>

proxy-prefilled
</details>

Clicking the button will revert to how it looks in the second screen shot (different from `f:multilineSecret` which has some extra UI around the text area that appears).

### All concealed by default

> Screen Shot 2019-04-19 at 01 22 31

### Changing one password

> Screen Shot 2019-04-19 at 01 22 37

### Proposed changelog entries

* Redesign password fields to prevent password auto-fill (undesirable except for the login form) and reduce the instances of browsers offering to update stored passwords. This change can be reverted any time by setting the system property `hudson.Functions.hidingPasswordFields` to `false`.

### Submitter checklist

- [n/a] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Comment by daniel-beck** [comment]
I want a CI build so I cannot keep this in draft 😭

**Comment by daniel-beck** [comment]
> the problem that browsers will autofill forms with username and password fields if they're empty

Even adding a password field long after the page is loaded (e.g. getting rid of the `choose`/`when` in the view) will autofill saved credentials in Firefox. 😭

**Comment by daniel-beck** [comment]
Right now I have something that appears to not crash & burn while preventing autofill or offers to save or update credentials in most cases.

Assuming of course, that Javascript is enabled.

How do we ensure this doesn't introduce worse problems than the one this is attempting to fix?

File: core/src/main/resources/lib/form/password.jelly
Comment (jvz) [inline core/src/main/resources/lib/form/password.jelly]:
```suggestion
                 name="${attrs.name ?: '_.'+attrs.field}"
```

File: core/src/main/resources/lib/form/password.jelly
Comment (jvz) [inline core/src/main/resources/lib/form/password.jelly]: Come to think of it, I think we don't need to inline this svg anymore. It was inlined in the standalone library because there was no good way to reference a static asset from a jar. This could be moved to a more appropriate location in jenkins to be a reusable static asset.

File: core/src/main/resources/lib/form/password.jelly
Comment (daniel-beck) [reply core/src/main/resources/lib/form/password.jelly]: Preferring a nicer `?w=1`

Comment (jvz) [reply core/src/main/resources/lib/form/password.jelly]: What?

Comment (daniel-beck) [reply core/src/main/resources/lib/form/password.jelly]: I thought this impacted the whitespace-ignoring diff but apparently it doesn't.

Comment (jvz) [reply core/src/main/resources/lib/form/password.jelly]: Ohh, ok.

**Comment by zbynek** [comment]
@daniel-beck to disable autocomplete in Chrome you can just add  `autocomplete="new-password"` to the input element. Hopefully this will get implemnted in FF in the future, given that it's mentioned in [MDN](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion)

**Comment by daniel-beck** [comment]
@zbynek I saw that on MDN, tried it, and failed. Thanks for linking the Firefox bug. Once that works and is reasonably widely supported (is Firefox ESR still around?) we can always change how we prevent autofill, but would rather not hold this change to wait for that (assuming it works otherwise).

**Comment by josephbrueggen** [comment]
@daniel-beck in the final screenshot (Changing one password) does selecting 'Build' beneath the form save the new password?

**Comment by daniel-beck** [comment]
@josephbrueggen No. It's just valid for the current build you're starting, behavior in general in unchanged from before.

**Comment by josephbrueggen** [comment]
@daniel-beck understood, thanks. This looks good to me.

**Review by Wadeck — COMMENTED** [review]
Tested with Chrome 74.0.3729.131, FireFox 66.0.3 (and 66.0.5), IE 11.706.17134.0, Edge 42.17134.1.0 under Windows 10.
Feature tested: 
- writing directly, 
- copy-pasting
- drag&dropping text

Every feature works as expected with all browsers.

My only concern was about the design, as in the multiline secret, the button is encapsulated in the border, with larger padding. But as Joe approved it, that's fine.

👍 for the proposal, but require to have tests to be approved :)

**Comment by daniel-beck** [comment]
I believe this to be done.

Six months without negative feedback on the basic approach means that, with the tests fixed, there cannot possibly be anything wrong here 😁

File: core/src/main/resources/lib/form/password.jelly
Comment (Wadeck) [inline core/src/main/resources/lib/form/password.jelly]: I think the name `value` is misleading here. Better use `passwordValue` or similar. Otherwise we could suppose it's just the plain text value, which is wrong as it's the encrypted value or `********`.

**Review by Wadeck — APPROVED** [review]
I expect this PR to break some ATH tests (assuming it was covered :p)

Same manual tests as before: 
Tested with Chrome 77.0.3865.90, FireFox 69.0.1 (and 69.0.2, thx auto updates) under Windows 10.

Required correction of previous tests means that the feature is already covered so for test wise we are good I imagine.

File: core/src/main/resources/lib/form/password.jelly
Comment (jvz) [inline core/src/main/resources/lib/form/password.jelly]: Nit: This SVG should really be put into the static files.

File: test/src/test/java/hudson/model/PasswordParameterDefinitionTest.java
Comment (jvz) [inline test/src/test/java/hudson/model/PasswordParameterDefinitionTest.java]: This should probably use `HtmlElementUtil.click()` instead. Same for the other one below.

**Review by jvz — APPROVED** [review]
Looks good overall. Left some nits.

**Review by jeffret-b — APPROVED** [review]
I've reviewed the code and the described behavior. That all looks good. I haven't tried testing it because it looks like it's pretty thoroughly tested.

It will be nice to get this behavior cleaned up. I've found it to be rather annoying at times.

**Comment by daniel-beck** [comment]
Anyone brave enough to write a `noscript` compatible variant of this the empty password field? I.e. only show it as a text field with the event handler if JavaScript is enabled? Perhaps @Wadeck has an idea how to do that?

**Comment by daniel-beck** [comment]
Given how broken the Jenkins UI is without JavaScript, we don't need to specifically take additional care of that.

**Comment by daniel-beck** [comment]
Manually tested that form field validation continues to work, and it will show on the "concealed" field when the page is loaded.

**Review by oleg-nenashev — APPROVED** [review]
Looks fine from the code standpoint, did not test manually.
I plan to merge it tomorrow if no negative feedback

**Comment by oleg-nenashev** [comment]
Ugh, I missed it during the reviews, because it was on the third page. Will go ahead and merge it

**Comment by oleg-nenashev** [comment]
@daniel-beck Just in case, did you have a chance to run ATH against it? The change is unlikely covered by smoke tests

**Review by oleg-nenashev — APPROVED** [review]
This change is likely to cause regressions in ATH. CC @olivergondza . I still think it worth merging even if we get a number of extra regressions there. I plan to merge it on Friday if no negative feedback

**Review by olivergondza — APPROVED** [review]
Agreed, if this is to improve things, we can handle som ATH breakage. We are looking for contributors, btw :P

**Comment by oleg-nenashev** [comment]
Maybe it should be a major RFE

**Comment by msmitty12** [comment]
How long would this change be expected to take before making it into an LTS release? We've had a lot of issues with a user's browser auto-filling forms in the Jenkins System Configuration without their knowledge and it seems like this change would address that.

**Comment by MarkEWaite** [comment]
It was first included in Jenkins 2.205.  Jenkins 2.204 was selected as the basis for the next long term support release.  Jenkins 2.204.1 will be released in January and will be the long term support base for 3 months.

An LTS release including this capability will most likely be available in April 2020.

**Comment by oleg-nenashev** [comment]
ETA for the new LTS baseline is Mar 2020, but indeed it will take a while
before it lands in LTS

**Comment by msmitty12** [comment]
I see, thanks for the update. Am I correct that this fix will avoid issues
with auto-fill (Chrome, LastPass, etc.) populating data in Jenkins system
configuration? Is there another workaround other than having every Jenkins
admin disable auto-fill in their applications?

**Comment by MarkEWaite** [comment]
@msmitty12 that matches with my reading of the [pull request description](https://github.com/jenkinsci/jenkins/pull/3991#issue-271866626).  Because it has released with Jenkins 2.205, you could test drive it yourself and verify the behavior with:

```
$ docker run --rm -i -t -p 8080:8080 jenkins/jenkins:2.205
```

That will provide you a locally running copy of Jenkins 2.205 that can be accessed on port 8080.

**Pull request merged**
